rfcs/experimental/dynamic_selection/README.md

```diff
@@ -0,0 +1,516 @@
+
+<img src="execution_flow.png" width=800>
+
+The free function `submit` receives a [Policy](#policy_req_id) object `p` and
+a function object `f` and returns an object `w` that can be waited on.
+A valid [Policy](#policy_req_id) must define a `submit` member function and a
+[Backend](#backend_req_id) must also define a `submit` member function. The
+requirements on [Policy](#policy_req_id) and [Backend](#backend_req_id) types
+are discussed later in this proposal. In the figure, the Policy's `submit` function
+selects a resource `r` and then passes `r`, an execution info handle `h`, and `f`
+to the Backend. The handle `h` is the mechanism for the Backend to report runtime
+information required by the Policy logic for making future selections, such as
+the execution time of a task when run on a specific resource.
+
+The [Backend](#backend_req_id) `submit` function invokes `f` with the selected
+resource as an argument and is also responsible for collecting and reporting
+any information required by the [Policy](#policy_req_id). The
+[Execution Info](#execution_info_id) section of this document describes how
+to use traits to determine what information needs to be reported. 
+
+## Named Requirements
+
+<a id="policy_req_id"></a>
+### Policy
+
+A Policy is an object with a valid dynamic selection heuristic.
+
+The type `T` satisfies *Policy* if given,
+
+- `p` an arbitrary identifier of type `T`
+- `args` an arbitrary parameter pack of types `typename… Args`
+- `s` a selection of a type `selection_t<T>` , which satisfies [Selection](#selection_req_id), and was made by `p`.
+- `f` a function object with signature `wait_t<T> fun(resource_t<T>, Args…);`
+- `r` a container of resources of type `std::vector<resource_t<T>`.
+
+| *Must* be well-formed | Description |
+| --------------------- | ----------- |
+| `p.get_resources()` | Returns a `std::vector<resource_t<T>>`. calling this function before `initialize` throws a `std::runtime_exception`. |
+| `p.select(args…)` | Returns `selection_t<T>` that satisfies [Selection](#selection_req_id). The selected resource must be within the set of resources returned by `p.get_resources()`. |
+| `p.submit(s, f, args…)` | Returns `submission_t<T>` that satisfies [Submission](#submission_req_id). The function invokes `f` but does not wait for the `wait_t<T>` object returned by it. |
+
+| Policy Traits* | Description |
+| ------- | ----------- |
+| `selection<T>::type`, `selection_t<T>` | The wrapped select type returned by `T`. Must satisfy [Selection](#selection_req_id). |
+| `resource<T>::type`, `resource_t<T>` | The backend defined resource type that is passed to the user function object. Calling `unwrap` an object of type `selection_t<T>` returns an object of type `resource_t<T>`. |
+| `wait_type<T>::type`, `wait_type_t<T>` | The backend type that is returned by the user function object. Calling `unwrap` on an object that satisfies [Submission](#submission_req_id) returns on object of type `wait_type_t<T>`. |
```

Comment:
The first entry in each of these rows seems wrong. As mentioned below
`policy_traits<T>::[selection,resource, wait]_type` exists for a policy `T`, but I don't know what `selection<T>::type` is, etc.  

It should either be ``policy_traits<T>::[selection,resource, wait]_type` as the first entry, or `[selection,resource, wait]_type` I think.

Reply:
Ok, good catch. I will fix these.